### ISSUE TITLE
Better explain delayed start

### ISSUE THREAD

**Issue by alextes** [ultrasoundmoney/oracle-client#19]
https://github.com/ultrasoundmoney/oracle-client/blob/09418e6a0264ac26db7752ac3a40c07e51a8ab8d/src/attestation_scheduler.rs#L149

Why do we end up in this situation if slots are run with a 24s timeout?

**Comment by ckoopmann** [comment]
@alextes So the reasoning behind this, is to avoid inconsistencies such that different clients might report prices from different timepoints in the 24s window.

The assumption is that we want all client instances to report the price at the beginning of the slot.
So even if we have a 24 second timeout it could happen that individual slots start executing a few seconds into the period. (for example when the maximum concurrency is reached and tasks get "queued"). In that case we want to immediately abort the task with the goal of getting back into "sync" of starting each execution at the beginning of the slot period.

I am not sure if this is still relevant though since your latest refactoring

**Comment by alextes** [comment]
> So even if we have a 24 second timeout it could happen that individual slots start executing a few seconds into the period.

How?

> (for example when the maximum concurrency is reached and tasks get "queued")

How could a task get "queued" if max concurrency is 2, tasks have a timeout of 24s, and slots are 12s each? Does the interval somehow start with a delay sometimes?

Might still be nice to leave it in anyway. It is far from obvious that were you to adjust concurrency down or timeout down, this definitely does become a concern.

**Comment by ckoopmann** [comment]
> > So even if we have a 24 second timeout it could happen that individual slots start executing a few seconds into the period.
> 
> How?
> 
> > (for example when the maximum concurrency is reached and tasks get "queued")
> 
> How could a task get "queued" if max concurrency is 2, tasks have a timeout of 24s, and slots are 12s each? Does the interval somehow start with a delay sometimes?
> 
> Might still be nice to leave it in anyway. It is far from obvious that were you to adjust concurrency down or timeout down, this definitely does become a concern.

Good point, with these settings it shouldn't happen to get "queued" in normal operation and probably won't need this check.

In addition to the risk  you mentioned of changing these values without keeping this restriction in mind there might also be another scenario:
The client running in an environment where it cannot actually run two tasks in parallel (single core hardware, etc.) - I don't know much about  parallel execution though, so only guessing here.

Either way if we want the oracle to refer to a very specific point in time (like the beginning of the block), we need to make sure that all clients actually get the price for that point in time. Otherwise we might run into very subtle issues with oracle accuracy, that are hard to debug. So it might be worth to be a bit extra careful (even slightly redundant in places)

Thinking about it more to ensure consistent timing of the price response it might be worth looking into the gopher response and see if it contains a timestamp that we could check.

**Comment by ckoopmann** [comment]
The gofer response does indeed include a timestamp which is probably a better reference to ensure correct timing of the price. 
However we might still want to keep the originally referenced check in place to abort a task when we are already "too late" and recover from the mentioned "queueing" situation if it ever occurs.


<img width="1436" alt="image" src="https://github.com/ultrasoundmoney/oracle-client/assets/15629702/f9caa79a-886c-4826-b377-cb0d5f78e894">
